### PR TITLE
Move DTLS code into the new write record layer

### PR DESCRIPTION
**Pull request by mattcaswell**
We move the DTLS write side into the new record layer architecture. This represents the last major component in the move to the new architecture. Some subsequent PRs are still required to resolve some loose ends and outstanding "TODO" items.

### CODE REVIEW

**Comment by paulidale** [comment]
CI failure looks relevant.

**Comment by mattcaswell** [comment]
> CI failure looks relevant.

Fixed (hopefully).

**Review by hlandau — APPROVED** [review]
I can't see anything at issue here.

What are we going to do about the disabled test? Can it be fixed in this PR, or does it need to be deferred?

**Comment by mattcaswell** [comment]
> What are we going to do about the disabled test? Can it be fixed in this PR, or does it need to be deferred?

This is one of the "loose ends" I plan to follow up on in subsequent PRs. There are a handful of "TODOs" left over in the code that need some attention - this being one of them. I decided to defer it because it wasn't strictly related to the DTLS code - its a TLSv1.3 test, which was missed in the earlier TLSv1.3 work.

File: ssl/record/rec_layer_s3.c [hunk 1]
```diff
@@ -54,8 +50,10 @@ void RECORD_LAYER_clear(RECORD_LAYER *rl)
 
 void RECORD_LAYER_release(RECORD_LAYER *rl)
 {
-    if (rl->numwpipes > 0)
-        ssl3_release_write_buffer(rl->s);
+    /*
+     * TODO(RECLAYER): Need a way to release the write buffers in the record
+     * layer on demand
+     */
```
Comment (levitte) [inline ssl/record/rec_layer_s3.c:56]: I did notice the change from a `RECORD_LAYER_get_wbuf()` call to direct memory allocation in `ssl/d1_lib.c`, I assume that this comment is kinda sorta connected to that change...

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
The CI failures are not relevant. Adding the "ready to merge" label.

**Comment by mattcaswell** [comment]
Pushed to master. Thanks for the reviews.

**Pull request closed without merging**